libyul/optimiser/BlockHasher.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+void BlockHasher::operator()(FunctionalInstruction const& _instr)
+{
+	hash64(compileTimeLiteralHash("FunctionalInstruction"));
+	hash8(static_cast<std::underlying_type_t<eth::Instruction>>(_instr.instruction));
+	hash64(_instr.arguments.size());
+	ASTWalker::operator()(_instr);
+}
+
+void BlockHasher::operator()(FunctionCall const& _funCall)
+{
+	hash64(compileTimeLiteralHash("FunctionCall"));
+	hash64(_funCall.functionName.name.hash());
+	hash64(_funCall.arguments.size());
+	ASTWalker::operator()(_funCall);
+}
+
+void BlockHasher::operator()(ExpressionStatement const& _statement)
+{
+	hash64(compileTimeLiteralHash("ExpressionStatement"));
+	ASTWalker::operator()(_statement);
+}
+
+void BlockHasher::operator()(Assignment const& _assignment)
+{
+	hash64(compileTimeLiteralHash("Assignment"));
+	hash64(_assignment.variableNames.size());
+	for (auto const& name: _assignment.variableNames)
+		(*this)(name);
+	visit(*_assignment.value);
+}
+
+void BlockHasher::operator()(VariableDeclaration const& _varDecl)
+{
+	hash64(compileTimeLiteralHash("VariableDeclaration"));
+	hash64(_varDecl.variables.size());
+	for (auto const& var: _varDecl.variables)
+	{
+		yulAssert(!m_variableReferences.count(var.name), "");
+		m_variableReferences[var.name] = VariableReference{
+			m_internalIdentifierCount++,
+			false
+		};
+	}
+	ASTWalker::operator()(_varDecl);
+}
+
+void BlockHasher::operator()(If const& _if)
+{
+	hash64(compileTimeLiteralHash("If"));
+	ASTWalker::operator()(_if);
+}
+
+void BlockHasher::operator()(Switch const& _switch)
+{
+	hash64(compileTimeLiteralHash("Switch"));
+	hash64(_switch.cases.size());
+	// Instead of sorting we could consider to combine
+	// the case hashes using a commutative operation here.
+	std::set<Case const*, SwitchCaseSorter> cases;
+	for (auto const& _case: _switch.cases)
+		cases.insert(&_case);
+
+	visit(*_switch.expression);
+	for (auto const& _case: cases)
+	{
+		if (_case->value)
+			(*this)(*_case->value);
+		(*this)(_case->body);
+	}
+}
+
+void BlockHasher::operator()(FunctionDefinition const& _funDef)
+{
+	hash64(compileTimeLiteralHash("FunctionDefinition"));
+	ASTWalker::operator()(_funDef);
+}
+
+void BlockHasher::operator()(ForLoop const& _loop)
+{
+	yulAssert(_loop.pre.statements.empty(), "");
+
+	hash64(compileTimeLiteralHash("ForLoop"));
+	ASTWalker::operator()(_loop);
+}
+
+void BlockHasher::operator()(Break const& _break)
+{
+	hash64(compileTimeLiteralHash("Break"));
+	ASTWalker::operator()(_break);
+}
+
+void BlockHasher::operator()(Continue const& _continue)
+{
+	hash64(compileTimeLiteralHash("Continue"));
+	ASTWalker::operator()(_continue);
+}
+
+
+void BlockHasher::operator()(Block const& _block)
+{
+	hash64(compileTimeLiteralHash("Block"));
+	hash64(_block.statements.size());
+	if (_block.statements.empty())
+		return;
+
+	BlockHasher subBlockHasher(m_blockHashes);
+	for (auto const& statement: _block.statements)
+		subBlockHasher.visit(statement);
+
+	m_blockHashes[&_block] = subBlockHasher.m_hash;
```

Comment:
Why is the hash of `_block` set without the external references?

Reply:
Well, the external references are hashed into the block hash when visiting identifiers, so in that sense they are part of the hash... for stuff like outlining it might make sense to store the external references alongside the hash right away, but since we don't need that for now, I kept it simple here.

Reply:
To explain why the external references are visited in the outer block, resp. hashed into the outer block, though: I'd think of a "subblock" similarly to a function acting on it's external references, so from the perspective of the outer block, I "call" the subblock with the variable names it references - note that for the outer block some of the external references in the inner block might be internal/local variables and some might still be external ones.

Reply:
Ah ok!